### ISSUE TITLE
dns.google is not compatible with "cloudflared"

### ISSUE THREAD

**Issue by bigdargon** [cloudflare/cloudflared#113]
Hi,
I am using debian and ubuntu operating systems, and I run cloudflared to use dns-over-https.

Previously I used https://dns.google.com/experimental address, everything works normally. But I switched to using a new address according to google's notification, I received an error message:
- If I use the address "https://dns.google/dns-query": level=error msg="failed to connect to an HTTPS backend \"https://dns.google/dns-query\"" error="returned status code 415"
- If I use the address "https://dns.google/resolve": level=error msg="failed to connect to an HTTPS backend \"https://dns.google/resolve\"" error="returned status code 400"
- If I use the address "https://8.8.8.8/dns-query": level=error msg="failed to connect to an HTTPS backend \"https://8.8.8.8/dns-query\"" error="returned status code 404"

What do I have to do to fix this problem? Thank you!

**Comment by bigdargon** [comment]
I created the issue here https://issuetracker.google.com/issues/136198937 and got the answer. Can Cloudflared update for compatibility?

> https://dns.google/dns-query is the right endpoint to use.
> 
> The problem is that (your version of) cloudflared is apparently still using the MIME type application/dns-udpwireformat rather than application/dns-message for the Content-Type header
> 
> From what I can tell, cloudflared is using the CoreDNS modular DNS server to implement its DNS to DoH proxy. The current version of CoreDNS DoH support (https://github.com/coredns/coredns/blob/master/plugin/pkg/doh/doh.go) uses application/dns-message, but the DoH support in CoreDNS was reconfigured and it is possible that cloudflared might need changes to work with the current CoreDNS architecture.
> 
> If you have been using the same version of cloudflared for a year or more, you should download a new copy and see if it still has this problem.
> 
> If the latest cloudflared still gives a 415 error, you might want to open an issue or request support on the Cloudflare community forum. If you are familiar with compiling Go programs, you could try to build cloudflared yourself and see what version of CoreDNS or the doh plugin it is using.
> 
> In the meantime, you can continue to use the https://dns.google.com/experimental endpoint, at least for another few weeks until you get the software issues sorted.

**Comment by rezaxdi** [comment]
Ok, due to today's cloudlfare outage suddenly I was not able to visit any site because dns was down. I tried to use google dns in cloudflared client and it was not working. So I had to uninstall cloudflared and get back to dnscrypt. I think cloudflared needs to update it's client to be compatible with other DoH servers.

**Comment by lib-xdown** [comment]
I think cloudflared needs to update it's client to be compatible with other DoH servers.

**Comment by sssilver** [comment]
We've merged https://github.com/cloudflare/cloudflared/pull/108, and will release it soon -- does this help resolve the problem?

**Comment by ndrwy** [comment]
Thanks @sssilver for looking into this, I've done some testing and below are my results:

```
# cloudflared --version
cloudflared version 2019.7.0 (built 2019-07-11-1656 UTC)
```

```
cloudflared[27174]: time="2019-07-11T22:48:42-04:00" level=error msg="failed to connect to an HTTPS backend \"https://dns.google/dns-query\"" error="failed to perform an HTTPS request: Post https://dns.google/dns-query: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)"
cloudflared[27316]: time="2019-07-11T22:50:11-04:00" level=error msg="failed to connect to an HTTPS backend \"https://dns.google/resolve\"" error="failed to perform an HTTPS request: Post https://dns.google/resolve: dial tcp: lookup dns.google on 127.0.0.1:53: read udp 127.0.0.1:53349->127.0.0.1:53: i/o timeout"
cloudflared[27256]: time="2019-07-11T22:49:01-04:00" level=error msg="failed to connect to an HTTPS backend \"https://8.8.8.8/dns-query\"" error="returned status code 404"
```

**Comment by bigdargon** [comment]
Thanks @sssilver but still not working!

**Comment by sssilver** [comment]
2019.7.0 is right. Can you confirm that the `Content-Type` is set to what's expected? If so, then the issue must be something else.

**Comment by lawliet89** [comment]
I am using 

```
pi@raspberrypi:~ $ cloudflared --version
cloudflared version 2019.9.0 (built 2019-09-06-0334 UTC)
```

When I set run cloudflared with `--port 5053 --upstream https://8.8.4.4/dns-query --upstream https://8.8.8.8/dns-query`, I get

```
Sep 08 10:04:21 raspberrypi systemd[1]: Started cloudflared DNS over HTTPS proxy.
Sep 08 10:04:21 raspberrypi cloudflared[17317]: time="2019-09-08T10:04:21+08:00" level=info msg="Adding DNS upstream" url="https://8.8.4.4/dns-query"
Sep 08 10:04:21 raspberrypi cloudflared[17317]: time="2019-09-08T10:04:21+08:00" level=info msg="Starting metrics server" addr="127.0.0.1:33133"
Sep 08 10:04:21 raspberrypi cloudflared[17317]: time="2019-09-08T10:04:21+08:00" level=info msg="Adding DNS upstream" url="https://8.8.8.8/dns-query"
Sep 08 10:04:21 raspberrypi cloudflared[17317]: time="2019-09-08T10:04:21+08:00" level=info msg="Starting DNS over HTTPS proxy server" addr="dns://localhost:5053"
Sep 08 10:04:56 raspberrypi cloudflared[17317]: time="2019-09-08T10:04:56+08:00" level=error msg="failed to connect to an HTTPS backend \"https://8.8.4.4/dns-query\"" error="returned status code 404"
Sep 08 10:04:56 raspberrypi cloudflared[17317]: time="2019-09-08T10:04:56+08:00" level=error msg="failed to connect to an HTTPS backend \"https://8.8.8.8/dns-query\"" error="returned status code 404"
Sep 08 10:05:06 raspberrypi cloudflared[17317]: time="2019-09-08T10:05:06+08:00" level=error msg="failed to connect to an HTTPS backend \"https://8.8.4.4/dns-query\"" error="returned status code 404"
Sep 08 10:05:06 raspberrypi cloudflared[17317]: time="2019-09-08T10:05:06+08:00" level=error msg="failed to connect to an HTTPS backend \"https://8.8.8.8/dns-query\"" error="returned status code 404"
Sep 08 10:05:13 raspberrypi cloudflared[17317]: time="2019-09-08T10:05:13+08:00" level=error msg="failed to connect to an HTTPS backend \"https://8.8.4.4/dns-query\"" error="returned status code 404"
Sep 08 10:05:13 raspberrypi cloudflared[17317]: time="2019-09-08T10:05:13+08:00" level=error msg="failed to connect to an HTTPS backend \"https://8.8.8.8/dns-query\"" error="returned status code 404"
```

**Comment by mnordhoff** [comment]
@lawliet89

Can you try using "https://dns.google/dns-query"?

I don't see `https://8.8.4.4/dns-query` or `https://8.8.8.8/dns-query` listed in Google's documentation.

**Comment by lawliet89** [comment]
Thanks @mnordhoff  It works with `https://dns.google/dns-query`.

I was just wondering how cloudflared was going to resolve `dns.google` in the first place so I decided to "play it safe" with the IP address.

**Comment by bigdargon** [comment]
I confirmed to **work** with `https://dns.google/dns-query`  in `cloudflared version 2019.9.0 (built 2019-09-06-0334 UTC)` (**not work** with `https://8.8.8.8/dns-query` & `https://8.8.4.4/dns-query`)

Thanks for the update!

**Comment by Tugzrida** [comment]
Just for future reference, this looks like an issue with Google's specific implementation.

Using some of the curl examples [here](https://developers.cloudflare.com/1.1.1.1/dns-over-https/wireformat/), the IP addresses do not work directly unless the `Host: dns.google` HTTP header is added. The [above-referenced docs](https://developers.google.com/speed/public-dns/docs/secure-transports#tls-sni) mention SNI, but not the Host header.

If someone wants to flag this with Google, then they should be accepting `Host: 8.8.8.8` and `Host: 8.8.4.4` for queries in addition to `Host: dns.google`.

**Comment by dpanic** [comment]
Hm, again not working with following opts set:

`CLOUDFLARED_OPTS=--port 5053 --upstream https://dns.google/dns-query`

**Comment by Tugzrida** [comment]
I've raised this with Google: https://issuetracker.google.com/issues/148296114

**Comment by m4niacjp** [comment]
Just tested today, dns.google **not** working with cloudflared.

**Comment by dpanic** [comment]
I see no update from Google guys...

**Comment by Tugzrida** [comment]
Yeah, I'm not sure how interested in fixing this Google is. A long term fix and general improvement to cloudflared would be allowing the manual setting of the HTTP Host header, which would fix this problem and allow #107 to be worked around too.

In the meantime, the way to get dns.google working is by setting your upstream to `https://dns.google/dns-query`. You may still run into problems with this if you've set cloudflared as your only system resolver as there is a bootstrapping paradox. To work around that you need to add `8.8.8.8 dns.google` to your hosts file.

**Comment by Tugzrida** [comment]
This was fixed by the introduction of bootstrapping in f18209a, which made it into v2020.4.0

dns.google can now be used simply by setting your upstream to `https://dns.google/dns-query`, no hosts file or other workaround needed.

Think this can be closed now.
